File: DevelopArchitecture/DevelopArchitecture/threeclass/photo/JPhotosPreviewsCell.m
//
//  JPhotosPreviewsCell.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 2019/4/2.
//  Copyright © 2019 Jeffrey. All rights reserved.
//

#import "JPhotosPreviewsCell.h"
#import "JPhotoPickerInterface.h"
#import <MediaPlayer/MediaPlayer.h>

@implementation JPhotosPreviewsCell{
    UIScrollView *m_BgScrollView;
     UIImageView *imageView;
    AVPlayer *player;
    AVPlayerLayer *playerLayer;
    UIButton *playButton;
    
    JPhotoPickerModel*m_data;
}
-(void)funj_addBaseCollectionView{

    
    imageView =[UIImageView funj_getImageView:self.bounds image:nil];
    imageView.contentMode = UIViewContentModeScaleAspectFit;
    if([JPhotosConfig share].m_currentIsVideo){
         [self.contentView addSubview:imageView];
        playButton =[UIButton funj_getButtons:self.bounds :nil  :JTextFCZero() :@[@"MMVideoPreviewPlay"] :self  :@"funj_selectToPlay:" :0 :nil];
         [self.contentView addSubview:playButton];
        [playButton funj_updateButtonSelectStyle:NO  :NO];
    }else{
        m_BgScrollView =[UIScrollView funj_getScrollView:self.bounds :nil];
        [self.contentView addSubview:m_BgScrollView];
        m_BgScrollView.bouncesZoom = YES;
        m_BgScrollView.maximumZoomScale = 2.5;
        m_BgScrollView.minimumZoomScale = 1.0;
        m_BgScrollView.multipleTouchEnabled = YES;
        m_BgScrollView.delegate = self;
        m_BgScrollView.scrollsToTop = NO;
        m_BgScrollView.showsHorizontalScrollIndicator = NO;
        m_BgScrollView.showsVerticalScrollIndicator = NO;
        m_BgScrollView.delaysContentTouches = NO;
        m_BgScrollView.canCancelContentTouches = YES;
        m_BgScrollView.alwaysBounceVertical = NO;
        [m_BgScrollView addSubview:imageView];
    }

    
 
}
- (nullable UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    return imageView;
}
- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
    
    CGFloat width = CGRectGetWidth(scrollView.frame);
    CGFloat height = CGRectGetHeight(scrollView.frame);
    
    CGFloat offsetX = (width > scrollView.contentSize.width) ? (width - scrollView.contentSize.width) * 0.5 : 0.0;
    CGFloat offsetY = (height > scrollView.contentSize.height) ? (height - scrollView.contentSize.height) * 0.5 : 0.0;
    imageView.center = CGPointMake(scrollView.contentSize.width * 0.5 + offsetX, scrollView.contentSize.height * 0.5 + offsetY);
}
-(void)funj_setBaseCollectionData:(JPhotoPickerModel*)data {
    m_data = data;
    [m_BgScrollView setZoomScale:1.0 animated:NO];
    LRWeakSelf(self);
    [JPhotoPickerInterface funj_getPhotoWithAsset:data.asset :PHImageRequestOptionsDeliveryModeOpportunistic photoWidth:imageView.width  completion:^(UIImage *photo, NSDictionary *info, BOOL isDegraded) {
        LRStrongSelf(self);
        self->imageView.image = photo;
    }];
    [self funj_stopAllPlayer];

}
-(void)funj_stopAllPlayer{
    if([JPhotosConfig share].m_currentIsVideo){
        [playerLayer removeFromSuperlayer];
        playerLayer = nil;
        [playButton setImage:[UIImage imageNamed:@"MMVideoPreviewPlay"] forState:UIControlStateNormal];
        imageView.hidden = NO;
        [player pause];
        player = nil;
    }
}


#pragma mark video
-(void)funj_selectToPlay:(UIButton*)sender{
    sender.selected = !sender.selected;
    if(sender.selected){
        if(playerLayer){
            [player play];
        }else{
            LRWeakSelf(self);
            [JPhotoPickerInterface funj_getVideoWithAsset:m_data.asset completion:^(AVPlayerItem * _Nullable item, NSDictionary * _Nullable dic) {
                LRStrongSelf(self);
                dispatch_async(dispatch_get_main_queue(), ^{
                    self->player = [AVPlayer playerWithPlayerItem:item];
                    self->playerLayer = [AVPlayerLayer playerLayerWithPlayer:self->player];
                    self->playerLayer.frame = self.bounds;
                    [self.layer insertSublayer:self->playerLayer atIndex:0];
                    [self->player play];
                    
                    self->imageView.hidden = YES;
                    [sender setImage:nil  forState:UIControlStateNormal];
                });
            }];
        }
    }else{
        [self funj_stopAllPlayer];
    }

 }

@end

File: DeveloperArchitecture/DeveloperArchitecture/threeclass/photo/cell/JPhotosPreviewsCell.swift
//
//  JPhotosPreviewsCell.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/11/7.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit
import MediaPlayer

class JPhotosPreviewsCell : JBaseCollectionViewCell, UIScrollViewDelegate {
    var m_BgScrollView : UIScrollView?
    var m_imageView : UIImageView?
    var m_player : AVPlayer?
    var m_playerLayer : AVPlayerLayer?
    var m_playButton : UIButton?
    var m_dataModel : JPhotoPickerModel?
    
    override func funj_addBaseCollectionView() {
        m_imageView = UIImageView(i: self.bounds, image: nil)
        m_imageView?.contentMode = .scaleAspectFit
        if JPhotosConfig.shared?.m_currentIsVideo ?? false {
            self.contentView.addSubview(m_imageView!)
            m_playButton = UIButton(i: self.bounds, title: nil, textFC: JTextFC())
                .funj_add(bgImageOrColor: ["MMVideoPreviewPlay"], isImage: true)
                .funj_add(targe: self, action: "funj_selectToPlay:", tag: 0)
                .funj_add(autoSelect: false)
        } else {
            m_BgScrollView = UIScrollView(i: self.bounds, delegate: nil)
            self.contentView.addSubview(m_BgScrollView!)
            m_BgScrollView!.bouncesZoom = true;
            m_BgScrollView!.maximumZoomScale = 2.5;
            m_BgScrollView!.minimumZoomScale = 1.0;
            m_BgScrollView!.isMultipleTouchEnabled = true;
            m_BgScrollView!.delegate = self;
            m_BgScrollView!.scrollsToTop = false;
            m_BgScrollView!.showsHorizontalScrollIndicator = false;
            m_BgScrollView!.showsVerticalScrollIndicator = false;
            m_BgScrollView!.delaysContentTouches = false;
            m_BgScrollView!.canCancelContentTouches = true;
            m_BgScrollView!.alwaysBounceVertical = false;
            m_BgScrollView?.addSubview(m_imageView!)
        }
    }
    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        return m_imageView
    }
    func scrollViewDidZoom(_ scrollView: UIScrollView) {
        let width = scrollView.width
        let height = scrollView.height
        let offsetX = width > scrollView.contentSize.width ? (width - scrollView.contentSize.width ) * 0.5 : 0.0
        let offsetY = height > scrollView.contentSize.height ? (height - scrollView.contentSize.height) * 0.5 : 0.0
        m_imageView?.center = CGPoint(x: scrollView.contentSize.width * 0.5 + offsetX, y: scrollView.contentSize.height * 0.5 + offsetY)
    }
    override func funj_setBaseCollectionData(_ data: Any) {
        m_dataModel = data as? JPhotoPickerModel
        m_BgScrollView?.setZoomScale(1.0, animated: false)
        JPhotoPickerInterface.funj_getPhotoWithAsset(phAsset: m_dataModel?.m_asset, deliveryMode: .opportunistic , width: m_imageView!.width) { [weak self] (image, dic, isDegraded) in
            self?.m_imageView?.image = image
        }
        self.funj_stopAllPlayer()
    }
    func funj_stopAllPlayer() {
        if JPhotosConfig.shared?.m_currentIsVideo ?? false {
            m_playerLayer?.removeFromSuperlayer()
            m_playerLayer = nil
            m_playButton?.setImage(UIImage(named: "MMVideoPreviewPlay"), for: .normal)
            m_imageView?.isHidden = false
            m_player?.pause()
            m_player = nil
        }
    }
    func funj_selectToPlay(_ sender : UIButton){
        sender.isSelected = !sender.isSelected
        if sender.isSelected {
            if m_playerLayer != nil {
                m_player?.play()
            }else {
                JPhotoPickerInterface.funj_getVideoWithAsset(asset: m_dataModel?.m_asset) { [weak self](item, dic) in
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {[weak self] in
                        self?.m_player = AVPlayer(playerItem: item)
                        self?.m_playerLayer = AVPlayerLayer(player: self?.m_player)
                        self?.m_playerLayer?.frame = self?.bounds ?? CGRectZero
                        if self?.m_playerLayer != nil {self?.layer.insertSublayer(self!.m_playerLayer!, at: 0)}
                        self?.m_player?.play()
                        
                        self?.m_imageView?.isHidden = true
                        sender.setImage(nil, for: .normal)
                    }
                }
            }
        } else {
            self.funj_stopAllPlayer()
        }
        
    }
}
